### PR TITLE
added erf op to math.py

### PR DESCRIPTION
**Pull request by sqali**
Hi @fchollet ,

Hope you are doing well. As discussed in issue keras-team/keras#18442, I have raised this PR to work on the ERF function. Kindly review the changes I have currently made. Once approved, I will go ahead with the implementation for JAX, torch, and numpy backends.

Thanks & Regards

### CODE REVIEW

File: keras_core/backend/torch/math.py [hunk 1]
```diff
@@ -408,3 +408,9 @@ def istft(
 def rsqrt(x):
     x = convert_to_tensor(x)
     return torch.rsqrt(x)
+
+
+def erf(x):
+    if not isinstance(x, torch.Tensor):
+        x = torch.tensor(x)
```
Comment (fchollet) [inline keras_core/backend/torch/math.py:415]: You can just do `x = convert_to_tensor(x)` unconditionally

File: keras_core/ops/math.py [hunk 2]
```diff
@@ -929,3 +929,41 @@ def rsqrt(x):
         return Rsqrt().symbolic_call(x)
     x = backend.convert_to_tensor(x)
     return backend.math.rsqrt(x)
+
+
+class Erf(Operation):
+    """Computes the error function of x element-wise.
+
+    Args:
+        input_tensor: A tensor of type `float32` or `float64`.
```
Comment (fchollet) [inline keras_core/ops/math.py:938]: It can have more types, no?

File: keras_core/ops/math.py [hunk 3]
```diff
@@ -929,3 +929,41 @@ def rsqrt(x):
...
+    def compute_output_spec(self, input_tensor):
+        return KerasTensor(shape=input_tensor.shape, dtype=input_tensor.dtype)
+
+    def call(self, input_tensor):
+        return backend.erf(input_tensor)
+
+
+@keras_core_export("keras_core.ops.erf")
+def erf(x):
+    """Functional interface to the `Erf` operation."""
```
Comment (fchollet) [inline keras_core/ops/math.py:968]: This is where the docstring should be, not the op above, since this is the public symbol.

File: keras_core/ops/math.py [hunk 4]
```diff
@@ -929,3 +929,41 @@ def rsqrt(x):
...
+    >>> y = Erf()(x)
+    # Using `float32` data type
+    >>> x_float32 = np.array([-3.0, -2.0], dtype=np.float32)
+    >>> y_float32 = Erf()(x_float32)
+    # Using large values
+    >>> x_large = np.array([1e10, -1e10])
+    >>> y_large = Erf()(x_large)
+    """
+
+    def __init__(self):
```
Comment (fchollet) [inline keras_core/ops/math.py:956]: Not needed

File: keras_core/ops/math.py [hunk 5]
```diff
@@ -929,3 +929,41 @@ def rsqrt(x):
...
+        super().__init__()
+
+    def compute_output_spec(self, input_tensor):
+        return KerasTensor(shape=input_tensor.shape, dtype=input_tensor.dtype)
+
+    def call(self, input_tensor):
```
Comment (fchollet) [inline keras_core/ops/math.py:962]: Just x

File: keras_core/ops/math_test.py [hunk 6]
```diff
@@ -835,3 +835,41 @@ def test_rsqrt(self):
...
+
+        # Output from the erf operation in keras_core
+        output_from_erf_op = kmath.erf(sample_values).numpy()
+
+        # Assert that the outputs are close
+        self.assertAllClose(expected_output, output_from_erf_op, atol=1e-5)
+
+    def test_erf_operation_dtype(self):
+        # Test for float32 and float64 data types
+        for dtype in [np.float32, np.float64]:
```
Comment (fchollet) [inline keras_core/ops/math_test.py:856]: Use strings for dtypes, and use tuples for static iterables rather than lists

**Review by fchollet — COMMENTED** [review]
Thanks for the PR! 👍

File: keras_core/backend/torch/math.py (hunk 1 above)
Comment (sqali) [reply keras_core/backend/torch/math.py]: Done

File: keras_core/ops/math.py (hunk 2 above)
Comment (sqali) [reply keras_core/ops/math.py]: Edited the comments based on that

File: keras_core/ops/math.py (hunk 3 above)
Comment (sqali) [reply keras_core/ops/math.py]: Relocated it!

File: keras_core/ops/math.py (hunk 4 above)
Comment (sqali) [reply keras_core/ops/math.py]: Removed it

File: keras_core/ops/math.py (hunk 5 above)
Comment (sqali) [reply keras_core/ops/math.py]: Replaced with x

File: keras_core/ops/math_test.py (hunk 6 above)
Comment (sqali) [reply keras_core/ops/math_test.py]: Done, used strings for dtypes and tuples for static iterables.

File: keras_core/ops/math.py [hunk 7]
```diff
@@ -929,3 +929,37 @@ def rsqrt(x):
...
     x = backend.convert_to_tensor(x)
     return backend.math.rsqrt(x)
+
+
+class Erf(Operation):
+    def compute_output_spec(self, x):
+        return KerasTensor(shape=x.shape, dtype=x.dtype)
+
+    def call(self, x):
+        return backend.erf(x)
```
Comment (fchollet) [inline keras_core/ops/math.py:939]: You need to call `backend.math`. Tests are failing.

File: keras_core/ops/math.py [hunk 8]
```diff
@@ -929,3 +929,37 @@ def rsqrt(x):
...
+
+    Args:
+        x: input tensor
+
+    Returns:
+        A tensor with the same type as `x`.
+
+    Examples:
+
+    # Basic usage
```
Comment (fchollet) [inline keras_core/ops/math.py:954]: Since you're not printing any outputs, just use a fenced code block for the code example.

**Review by fchollet — COMMENTED** [review]
Thanks for the updates!

File: keras_core/ops/math.py (hunk 7 above)
Comment (sqali) [reply keras_core/ops/math.py]: implemented it

File: keras_core/ops/math.py (hunk 8 above)
Comment (sqali) [reply keras_core/ops/math.py]: Done

**Comment by sqali** [comment]
Hi @fchollet ,

Thanks so much for the guidance. The Tests are failing because the errors are greater than the tolerance level (1×10^-5) set as you can see below.

| Metric                   | Value        |
|--------------------------|--------------|
| Max absolute difference  | 0.12837633   |
| Max relative difference  | 0.12837917   |


Kindly guide me on how to resolve this issue.

Thanks & Regards

**Comment by fchollet** [comment]
You can just lower the precision of this particular check to 1e-4.

**Comment by sqali** [comment]
Hi @fchollet ,

Thanks for the reply, Lowered the precision to 1e-4 such that it would resolve the issue but the maximum difference is around 0.12 which is way higher than 0.0001. Kindly advise.

File: keras_core/ops/math_test.py [hunk 9]
```diff
@@ -835,3 +835,41 @@ def test_rsqrt(self):
...
+            )
+            expected_output = (2 / np.sqrt(np.pi)) * np.vectorize(math.erf)(
+                sample_values
+            )
+            output_from_erf_op = kmath.erf(sample_values)
+            self.assertAllClose(expected_output, output_from_erf_op, atol=1e-4)
+
+    def test_erf_operation_edge_cases(self):
+        # Test for edge cases
+        edge_values = np.array([1e10, -1e10, 1e-10, -1e-10], dtype=np.float64)
```
Comment (fchollet) [inline keras_core/ops/math_test.py:868]: Your test values are too large. Try 1e5. This the source of the large discrepancy IMO.

Comment (sqali) [reply keras_core/ops/math_test.py]: I have implemented the changes, but I can see from the tests that it is failing for the below array examples.

- x: array([ 1.128379e+00, -1.128379e+00,  1.273240e-05, -1.273240e-05])
- x: array([-1.128354, -1.123101, -0.950886,  0. , 0.950886,  1.123101, 1.128354])

**Comment by fchollet** [comment]
Keras Core is becoming Keras 3, and we're switching development to the main repository! Please reopen this PR in the [keras-team/keras](https://github.com/keras-team/keras) repository. Unfortunately we aren't able to automatically transfer PRs (but we have transferred all issues).
